### PR TITLE
feat(cardano): native-token support + manual-CBOR signing

### PR DESCRIPTION
**Pull request by gastonm5**
## Summary

Brings vultisig-ios to parity with vultisig-sdk #238 (Cardano native-token discovery, balance, metadata, TokenBundle in signing input) and #239 (manual-CBOR envelope bypassing WalletCore's broken Cardano compiler). Closes vultisig-ios#4137 / vultisig-windows#2870 parity.

Plan + design decisions: [`wiki/pages/projects/vultisig/cardano-native-tokens/spec/`](https://github.com/gastonm5/wiki/tree/main/pages/projects/vultisig/cardano-native-tokens) (private wiki).

### Signing refactor (highest-risk change — review first)

- Drop `TransactionCompiler.compileWithSignatures(...)` and ~180 LOC of post-compile CBOR parsing (`extractCardanoTransactionBody` / `findEndOfCBORItem` / `readCBORLength` / `calculateCardanoTransactionHash`).
- Hand-build `[body, witnessSet, true, null]` per the SDK byte recipe in `CardanoSignedTxBuilder`. **The body bytes are embedded verbatim — never re-encoded** — so the Ed25519 signature (Blake2b-256 of those bytes) stays valid. Witness set is hand-encoded CBOR: `a1 00 81 82 5820<vkey> 5840<sig>`.
- `txId = preSigningOutput.dataHash.hexString` — the same hash MPC signed. Threaded through `broadcastTransaction(signedTransaction:precomputedTxId:)` so the 3117 \"already in mempool\" fallback uses the precomputed hash; no more parser.

### Native-token model

- `CardanoAssetId` — `policy_id.asset_name` parser/builder (hex, lowercased, length-validated). Mirrors SDK `cardanoAssetId.ts`.
- `CardanoNativeTokensService` (actor) — Koios `address_assets` for discovery, `address_assets`/`asset_info` for resolve. Ticker fallback chain matches SDK exactly: `registry.ticker` → `asset_name_ascii` → first 8 hex chars upper. `asset_info` cached 7 days.
- `CardanoExtendedUtxo` + extended UTXO fetch (Koios `_extended: true`) with deterministic `(hash, index)` ordering so MPC peers reach identical body bytes through the WalletCore planner.
- `CardanoService.getBalance(coin:address:)` dispatches on `contractAddress` — empty → ADA via `address_info`, non-empty → matching asset quantity via `address_assets`.
- TokenBundle populated on the WalletCore `Transfer` message when `contractAddress` is non-empty; amount encoded as minimal big-endian unsigned bytes (zero → single `0x00` to match SDK behaviour).

### UI (generic-with-branches — mirrors Windows; no Cardano-only screens)

Windows research confirmed there are zero Cardano-specific UI files in vultisig-windows; every CNT touchpoint is a single chain check at the validation/service layer riding the existing `AddCustomTokenPage` and Send flow. iOS does the same.

- **`CustomTokenScreen`** — Cardano branch in `fetchTokenInfo` calls `CardanoNativeTokensService.shared.resolveMetadata(assetId:)`. Validation uses `CardanoAssetId.parse` for Cardano, `AddressService.validateAddress` for every other chain. Maps `assetNotFound` to the existing `TokenNotFoundError` so the existing `ActionBannerView` retry pattern handles it.
- **Auto-discovery** — new Phase-4 hook in `BalanceService.updateBalances` mirrors Windows' `CoinFinder` (silent, automatic). Filters tokens already in `vault.coins` by lowercased `contractAddress` and persists new ones via `CoinService.addToChain`.
- **Min-ADA Verify error** — `SendCryptoVerifyLogic.validateBalanceWithFee` now checks vault ADA balance ≥ `fee + 2 × CardanoHelper.defaultMinUTXOValue` for CNT sends and surfaces `cardanoOutputBelowMinAda` through the existing alert pattern (same path as the BTC dust-check error).

### Why no `commondata` change

Matches SDK + Windows: each MPC peer re-fetches extended UTXOs from Koios at sign-time. No per-UTXO assets on the wire; `UtxoInfo` stays lovelace-only as today.

### Localisation

`cardanoOutputBelowMinAda` added to all 8 locales (en/de/es/hr/it/ko/pt/zh-Hans). `sort_localizable.py` ran clean. Crypto terms (ADA, UTXO, Cardano) kept in English per repo convention.

## Test plan

- [x] 51 new unit tests across 5 test files, byte-parity-tested where applicable:
  - `CardanoAssetIdTests` (12) — round-trip, hex/length validation, lowercasing
  - `CardanoExtendedUtxoTests` (9) — Koios JSON decode, asset mapping, edge cases (null asset_list / decimals / non-numeric quantity)
  - `CardanoNativeTokensServiceTests` (10) — ticker derivation, hex→ASCII (incl. 7-bit mask), policy-prefix fallback, lowercasing
  - `CardanoSignedTxBuilderTests` (12) — full envelope parity, `cborBytes` boundary tests at len 23/24/255/256/65535, length-validation throws
  - `CardanoTokenBundleTests` (7) — bundle construction, big-endian amount encoding (incl. >UInt64 and zero), invalid-asset-id throws
- [x] `ChainHelperTests` + `UTXOChainsHelperTest` regression suites still green
- [x] SwiftLint: 0 violations across all touched files
- [x] `xcodebuild build` clean
- [ ] **Mainnet smoke** — small-amount ADA-only send verifies the manual-CBOR envelope on the largest cohort (do this BEFORE moving out of draft)
- [ ] **Mainnet CNT send** — small amount of a known native token (e.g. USDM) on FastVault: receiver balance updates, ADA fee deducted, txid resolves on explorer
- [ ] **Custom-token add** — paste a `policy_id.asset_name` for an unknown CNT, verify metadata preview + "Add" persistence
- [ ] **Auto-discovery** — load a vault with held CNT, confirm tokens appear silently after first balance refresh
- [ ] **Min-ADA error** — vault with CNT but \<2.8 ADA tries to send; alert appears on Verify

## Out of scope

- Cardano dApp signing (CIP-30) — separate workstream, will reuse the commondata `SignBitcoin` PSBT pattern when it lands.
- Multi-CNT in a single transfer — Windows + SDK do single-token transfers; we match that.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Full Cardano native token support: asset-id parsing/validation, token metadata resolution, extended UTXO discovery, token-aware balances, automatic token discovery/add-to-vault, Cardano token selection, Cardano-specific signing payload handling, and a new signed-transaction builder.

* **Bug Fixes / Validation**
  * Enforced Cardano minimum-ADA checks for token sends with clear insufficient-ADA error.

* **Localization**
  * Added "insufficient ADA / minimum UTXO" message in multiple languages.

* **Tests**
  * Extensive unit tests covering parsing, UTXO mapping, metadata resolution, signing/builder behavior, and message handling.

### CODE REVIEW

**Comment by gastonm5** [comment]
Addressed CodeRabbit feedback in `0999f0e9d`. Mapping (with rationale where I diverged from the suggestion):

| Comment | Status | Notes |
|---|---|---|
| `CardanoAssetId.swift:47` — odd-length asset-name | **Fixed** | New `oddAssetNameLength(Int)` error case; assert covered by `testParseRejectsOddLengthAssetName` |
| `CardanoExtendedUtxo.swift:41` — malformed asset silently dropped | **Fixed** | Whole UTxO now returns `nil` if any asset fails to parse; also rejects negative quantities. Two new test cases |
| `CardanoNativeTokensService.swift:82` — empty assetName ticker | **Fixed** | Extra fallback to `policyId.prefix(8).uppercased()` after the existing chain, so `resolveMetadata` agrees with `discoverTokens` for nameless CNTs |
| `CardanoNativeTokensService.swift:119` — `hexToAscii` masking | **N/A** | Function lives at `StringExtension.swift:48` after the f94fa0d move; addressed via the StringExtension comment below |
| `CardanoSignedTxBuilder.swift:33` — 64 KiB body limit | **Fixed** | Removed `maxBodyLength` + `bodyTooLarge` error case. Replaced the rejection test with `testAcceptsLargeTxBody` (70 KiB body) |
| `zh-Hans:131` — untranslated `vault` | **Fixed** | `vault` → `保险库` (matches the other ~40 vault-prefixed entries in this locale) |
| `StringExtension.swift:61` — `hexToAscii` partial-decode + 7-bit mask | **Fixed** | Strict odd-length rejection (was partial-decoding 3-char inputs); printable-ASCII-only (0x20–0x7E). Updated `testHexToAsciiRejectsNonPrintableBytes` (replaces the prior `testHexToAsciiHighBitMaskedToSeven`) and added the multi-byte `"414"` odd-length case. Note: this is a deliberate deviation from the SDK's `Buffer.toString('ascii')` byte-mask — for non-ASCII asset names the user-visible result is the same (callers fall back to the policy prefix), but iOS no longer fabricates ASCII characters from high-bit bytes |
| `Cardano.swift:187` (out-of-diff) — propagate per-UTXO assets to `TW_Cardano_Proto_TxInput.tokenAmount` | **Skipping** | The SDK signing-input resolver doesn't populate per-UTXO assets either — `vultisig-sdk/.../signingInputs/resolvers/cardano.ts:54-66` builds lovelace-only `TxInput`s and only sets the `TokenBundle` on the `Transfer` output. The user's directive on this PR was to match SDK + Windows behaviour exactly; intentionally diverging on this is a separate workstream. If WalletCore's planner is broken on inputs without asset data, the SDK has the same bug. Documented in the wiki spec under §6 (no-op rationale). Happy to revisit if you have a counter-example where the SDK works and iOS fails |

51 → **57** unit tests now (replaced the byte-mask test, added the odd-length / non-printable / oddAssetNameLength / negative-quantity / large-body cases). ChainHelperTests + UTXOChainsHelperTest regressions still green. SwiftLint 0/0.

File: VultisigApp/VultisigApp/Blockchain/Cardano/Service/CardanoNativeTokensService.swift [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    var bytes: [UInt8] = []
+    var iterator = hex.makeIterator()
+    while let high = iterator.next(), let low = iterator.next() {
+        guard
+            let highValue = high.hexDigitValue,
+            let lowValue = low.hexDigitValue
+        else { return "" }
+        bytes.append(UInt8(highValue * 16 + lowValue) & 0x7F)
+    }
+    return String(bytes: bytes, encoding: .ascii) ?? ""
```
[the remark this answers is not in the document]
Comment (gastonm5) [reply VultisigApp/VultisigApp/Blockchain/Cardano/Service/CardanoNativeTokensService.swift]: Declining this one — the `& 0x7F` mask is intentional and documented.

`hexToAscii()` was moved to `StringExtension.swift` in 4e007f001; the docstring there spells out why we keep the mask:

> Decode a hex string into ASCII characters, matching Node's `Buffer.from(hex, 'hex').toString('ascii')` 7-bit mask but stripping non-printable bytes (outside `0x20…0x7E`) so the UI doesn't render control characters as garbage. Returns `""` on odd length, non-hex input, or when every byte is non-printable.
>
> Used for Cardano native-token ticker derivation. CIP-68 fungible tokens (label `(333)` = `0x0014df10`) and CIP-67 reference tokens prepend a 4-byte binary label to the asset name; stripping the non-printable prefix surfaces the readable ticker (e.g. `0014df105553444d` → `_USDM`).

Concretely, for USDM (mainnet asset id `c48cbb…ad.0014df105553444d`):

- `0x00 0x14` mask to `0x00 0x14`, both filtered out (not printable) ✅
- `0xDF` masks to `0x5F` = `_`, passes the printable filter → ends up as a leading underscore
- `0x10` masks to `0x10`, filtered out ✅
- `5553444d` = `USDM` ✅

Final ticker: `_USDM`. The leading underscore is a known artifact of `& 0x7F` collapsing `0xDF` (the third byte of the CIP-67 (333) prefix) into a printable ASCII slot, but the team accepted it because the resulting ticker is still recognisable.

Removing the mask (CodeRabbit's suggestion) and rejecting any byte > `0x7F` would make the `0xDF` byte fail the whole function. `hexToAscii()` would return `""`, the caller in `CardanoNativeTokensService.makeMetadata` would fall back to `String(policyId.prefix(8)).uppercased()`, and USDM (and every other CIP-67/68 token) would render as `C48CBB3D` instead of `_USDM`. That's a worse UX than the current cosmetic underscore.

Long-term, the cleaner fix is to detect the CIP-67/68 label prefix (first 4 bytes when the leading byte is `0x00`) and strip it explicitly before decoding the rest as ASCII — that's a separate refactor that would let us drop the mask entirely. Tracking that as a follow-up rather than regressing the v1 ticker derivation here.

**Comment by johnnyluo** [comment]
@gastonm5 , The PR in commondata has been merged, can you please update this PR to reference to the latest on main of commondata?

**Comment by gastonm5** [comment]
> @gastonm5 , The PR in commondata has been merged, can you please update this PR to reference to the latest on main of commondata?

Updated @johnnyluo

**Review by johnnyluo — APPROVED** [review]
LGTM
